File: python/814. Binary Tree Pruning.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pruneTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not root: return root
        root.left = self.pruneTree(root.left)
        root.right = self.pruneTree(root.right)

        if not root.left and not root.right:
            return None if root.val==0 else root
    
        return root

File: go/src/leetcode/814/814. Binary Tree Pruning.go
/**
@author ZhengHao Lou
Date    2022/07/21
*/
package main

/**
https://leetcode.cn/problems/binary-tree-pruning/
思路：后续遍历
*/
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pruneTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	if root.Val == 0 && root.Left == nil && root.Right == nil {
		return nil
	}

	root.Left = pruneTree(root.Left)
	root.Right = pruneTree(root.Right)

	if root.Val == 0 && root.Left == nil && root.Right == nil {
		return nil
	}

	return root
}
